src/btree/bt_ovfl.c

```diff
@@ -109,51 +121,100 @@ __ovfl_cache_col_visible(
  *	row-store: check for a globally visible update.
  */
 static bool
-__ovfl_cache_row_visible(WT_SESSION_IMPL *session, WT_PAGE *page, WT_ROW *rip)
+__ovfl_cache_row_visible(WT_SESSION_IMPL *session, WT_UPDATE *upd)
 {
-	WT_UPDATE *upd;
-
 	/* Check to see if there's a globally visible update. */
-	for (upd = WT_ROW_UPDATE(page, rip); upd != NULL; upd = upd->next)
-		if (__wt_txn_upd_visible_all(session, upd))
+	for (; upd != NULL; upd = upd->next)
+		 if ((upd->type == WT_UPDATE_STANDARD ||
+		     upd->type == WT_UPDATE_DELETED) &&
+		     __wt_txn_upd_visible_all(session, upd))
 			return (true);
 
 	return (false);
 }
 
 /*
- * __ovfl_cache --
- *	Cache a deleted overflow value.
+ * __ovfl_cache_append_update --
+ *	Append an overflow value to the update list.
  */
 static int
-__ovfl_cache(WT_SESSION_IMPL *session, WT_PAGE *page, WT_CELL_UNPACK *unpack)
+__ovfl_cache_append_update(WT_SESSION_IMPL *session, WT_PAGE *page,
+    WT_UPDATE *upd_list, WT_CELL_UNPACK *unpack, WT_UPDATE **updp)
 {
 	WT_DECL_ITEM(tmp);
 	WT_DECL_RET;
-	size_t addr_size;
-	const uint8_t *addr;
+	WT_UPDATE *append, *upd;
+	size_t size;
 
-	addr = unpack->data;
-	addr_size = unpack->size;
+	*updp = NULL;
 
+	/* Read the overflow value. */
 	WT_RET(__wt_scr_alloc(session, 1024, &tmp));
+	WT_ERR(__ovfl_read(session, unpack->data, unpack->size, tmp));
+
+	/*
+	 * Create an update entry with an impossibly low transaction ID to
+	 * ensure global visibility, append it to the update list.
+	 *
+	 * We don't need locks or barriers in this function: any thread reading
```

Comment:
I thought we only came through here via reconciliation, so the page should be in `WT_REF_LOCKED`, am I missing a path?